cpp/velox/shuffle/VeloxShuffleWriter.cc

```diff
@@ -1420,26 +1398,31 @@ arrow::Status VeloxShuffleWriter::splitFixedWidthValueBuffer(const facebook::vel
       *actual = 0;
       return arrow::Status::OK();
     }
-    EvictGuard{splitState_};
+    auto originalState = splitState_;
+    splitState_ = SplitState::kUnevictable;
 
     int64_t reclaimed = 0;
-    if (reclaimed < size && shrinkPartitionBuffersBeforeSpill()) {
+    if (reclaimed < size && shrinkPartitionBuffersBeforeSpill(originalState)) {
       ARROW_ASSIGN_OR_RAISE(auto shrunken, shrinkPartitionBuffers());
       reclaimed += shrunken;
     }
     if (reclaimed < size) {
       ARROW_ASSIGN_OR_RAISE(auto cached, evictCachedPayload());
       reclaimed += cached;
     }
-    if (reclaimed < size && shrinkPartitionBuffersAfterSpill()) {
-      ARROW_ASSIGN_OR_RAISE(auto shrunken, shrinkPartitionBuffersMinSize(size - reclaimed));
+    if (reclaimed < size && shrinkPartitionBuffersAfterSpill(originalState)) {
+      ARROW_ASSIGN_OR_RAISE(auto shrunken, shrinkPartitionBuffersMinSize(originalState, size - reclaimed));
       reclaimed += shrunken;
     }
-    if (reclaimed < size && evictPartitionBuffersAfterSpill()) {
+    if (reclaimed < size && evictPartitionBuffersAfterSpill(originalState)) {
       ARROW_ASSIGN_OR_RAISE(auto evicted, evictPartitionBuffersMinSize(size - reclaimed));
       reclaimed += evicted;
     }
     *actual = reclaimed;
+
+    // Restore the state.
+    splitState_ = originalState;
+
```

Comment:
I think the previous way is good since it took exception throwing into account.

Probably we just need for a independent `EvictState` which is specialized to evicting flow control.